crates/runtime-table-partition/src/provider.rs

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{any::Any, collections::HashMap, sync::Arc};
+
+use arrow_schema::SchemaRef;
+use async_trait::async_trait;
+use datafusion::{
+    catalog::{Session, TableProvider},
+    common::Constraints,
+    datasource::TableType,
+    error::DataFusionError,
+    physical_plan::ExecutionPlan,
+    prelude::Expr,
+};
+
+use crate::{Error, creator::PartitionCreator};
+
+type ScalarValueString = String;
+
+#[derive(Debug)]
+pub struct PartitionTableProvider {
+    _creator: Arc<dyn PartitionCreator>,
+    _partition_by: Vec<Expr>,
+    _partitions: HashMap<ScalarValueString, Arc<dyn TableProvider>>,
+    schema: SchemaRef,
+}
+
+impl PartitionTableProvider {
+    /// Create a new [`PartitionTableProvider`] and attempt to infer existing
+    /// `partitions` using the specified `creator`.
+    ///
+    /// # Errors
+    /// Returns an error if partition inferencing fails.
+    pub async fn new(
+        creator: Arc<dyn PartitionCreator>,
+        partition_by: Vec<Expr>,
+        schema: SchemaRef,
+    ) -> Result<Self, Error> {
+        let partitions = creator
+            .infer_existing_partitions()
```

Comment:
We should have validation here that the inferred partitions actually match the valid output of the expression. Ideally the actual possible values, but at least the data type.